File: solver/generate_levels.py
#!/usr/bin/env pypy3
"""
Generate level JSON files for the laser puzzle game.
Computes optimal scores using beam search and outputs to solver/levels/

Usage:
    python generate_levels.py                    # Generate 7 days (today and 6 days back)
    python generate_levels.py --days 30          # Generate 30 days
    python generate_levels.py --date 2026-02-01  # Generate specific date
    python generate_levels.py --start 2026-01-29 --end 2026-02-28  # Generate date range
"""

import argparse
import json
import os
from datetime import datetime, timedelta
from pathlib import Path

from puzzles import PUZZLES, ORIGINAL_PUZZLE_COUNT, PuzzleConfig
from used_puzzles import USED_PUZZLES
from simulator import beam_search_solver

# Date ranges for puzzle sets
# Original puzzles: Jan 22-28, 2026 (indices 0-6 in USED_PUZZLES)
# Extended puzzles: Jan 29 - Feb 28, 2026 (indices 7-37 in USED_PUZZLES)
# New puzzles: Mar 1, 2026+ (in PUZZLES)
ORIGINAL_START = datetime(2026, 1, 22)
EXTENDED_START = datetime(2026, 1, 29)
NEW_PUZZLES_START = datetime(2026, 3, 1)


def get_puzzle_for_date(date: datetime) -> PuzzleConfig:
    """
    Get the puzzle configuration for a given date.

    Date ranges:
    - Jan 22-28, 2026: Original puzzles (USED_PUZZLES indices 0-6)
    - Jan 29 - Feb 28, 2026: Extended puzzles (USED_PUZZLES indices 7-37)
    - Mar 1, 2026+: New puzzles (PUZZLES)
    """
    if date < EXTENDED_START:
        # Original behavior for Jan 22-28
        index = date.timetuple().tm_yday % ORIGINAL_PUZZLE_COUNT
        return USED_PUZZLES[index]
    elif date < NEW_PUZZLES_START:
        # Extended puzzles for Jan 29 - Feb 28
        days_since_extended = (date - EXTENDED_START).days
        extended_count = len(USED_PUZZLES) - ORIGINAL_PUZZLE_COUNT
        index = ORIGINAL_PUZZLE_COUNT + (days_since_extended % extended_count)
        return USED_PUZZLES[index]
    else:
        # New puzzles for Mar 1+
        if not PUZZLES:
            raise ValueError(
                f"No new puzzles available for {date.strftime('%Y-%m-%d')}. "
                "Add new puzzles to puzzles.py following the documented rules."
            )
        days_since_new = (date - NEW_PUZZLES_START).days
        index = days_since_new % len(PUZZLES)
        return PUZZLES[index]


def generate_level(date_str: str, config: PuzzleConfig) -> dict:
    """Generate a level for the given date using the provided puzzle config."""
    print(f"Computing optimal score for {date_str} ({config.name})...")
    result = beam_search_solver(config, verbose=False)
    optimal_score = result['path_length']
    print(f"  Optimal score: {optimal_score}")

    return {
        "date": date_str,
        "gridWidth": config.width,
        "gridHeight": config.height,
        "laserConfig": {
            "x": config.laser_x,
            "y": config.laser_y,
            "direction": config.laser_dir.to_string(),
        },
        "obstacles": [{"x": x, "y": y} for x, y in config.obstacles],
        "mirrorsAvailable": config.num_mirrors,
        "optimalScore": optimal_score,
    }


def generate_for_date(date: datetime, output_dir: Path):
    """Generate and save a level for a specific date."""
    date_str = date.strftime("%Y-%m-%d")
    config = get_puzzle_for_date(date)
    level = generate_level(date_str, config)

    output_path = output_dir / f"{date_str}.json"
    with open(output_path, "w") as f:
        json.dump(level, f, indent=2)
    print(f"  Saved to {output_path}")


def main():
    parser = argparse.ArgumentParser(description="Generate level JSON files")
    parser.add_argument(
        "--days",
        type=int,
        default=7,
        help="Number of days to generate (default: 7, going back from today)",
    )
    parser.add_argument(
        "--date",
        type=str,
        help="Generate a specific date (YYYY-MM-DD format)",
    )
    parser.add_argument(
        "--start",
        type=str,
        help="Start date for range generation (YYYY-MM-DD format)",
    )
    parser.add_argument(
        "--end",
        type=str,
        help="End date for range generation (YYYY-MM-DD format)",
    )
    args = parser.parse_args()

    # Create output directory
    output_dir = Path(__file__).parent / "levels"
    output_dir.mkdir(exist_ok=True)

    if args.start and args.end:
        # Generate date range
        start_date = datetime.strptime(args.start, "%Y-%m-%d")
        end_date = datetime.strptime(args.end, "%Y-%m-%d")
        current = start_date
        while current <= end_date:
            generate_for_date(current, output_dir)
            current += timedelta(days=1)
    elif args.date:
        # Generate single specific date
        date = datetime.strptime(args.date, "%Y-%m-%d")
        generate_for_date(date, output_dir)
    else:
        # Generate multiple days going back from today
        today = datetime.now()

        for i in range(args.days - 1, -1, -1):
            date = today - timedelta(days=i)
            generate_for_date(date, output_dir)

    print("\nDone!")


if __name__ == "__main__":
    main()

File: solver/generate_levels.ts
#!/usr/bin/env node
'use strict';
/**
 * Generate level JSON files for the laser puzzle game.
 * Computes optimal scores using beam search and outputs to solver/levels/
 *
 * Usage:
 *   npx tsx generate_levels.ts --date 2026-03-01           # Generate specific date
 *   npx tsx generate_levels.ts --start 2026-03-01 --end 2026-03-07  # Generate date range
 *   npx tsx generate_levels.ts --beam-width 2000 --workers 4  # Tuning options
 */

const fs = require('fs');
const path = require('path');
const os = require('os');

const { PUZZLES } = require('./puzzles');
const { solvePuzzle } = require('./solve');

// ── Helpers ───────────────────────────────────────────────────────────────────

const MAX_SPLITTERS = 2;

interface SolverOpts {
  beamWidth: number;
  workers: number;
}

interface PuzzleConfig {
  name: string;
  width: number;
  height: number;
  laserX: number;
  laserY: number;
  laserDir: number;
  obstacles: [number, number][];
  numMirrors: number;
  splitters?: [number, number, string][];
}

/**
 * Get the puzzle configuration for a given date.
 * Throws if no puzzle exists for that date.
 */
function getPuzzleForDate(dateStr: string): PuzzleConfig {
  if (!(dateStr in PUZZLES)) {
    throw new Error(
      `No puzzle configured for ${dateStr}. ` +
      'Add a new JSON file to solver/puzzles/ with this date.'
    );
  }
  return PUZZLES[dateStr];
}

/**
 * Generate a level for the given date using the provided puzzle config.
 * Returns the level object (same schema as Python generate_levels.py).
 */
async function generateLevel(dateStr: string, config: PuzzleConfig, solverOpts: SolverOpts) {
  const numSplitters = (config.splitters || []).length;
  if (numSplitters > MAX_SPLITTERS) {
    throw new Error(
      `${dateStr}: puzzle has ${numSplitters} splitters (max ${MAX_SPLITTERS}). ` +
      'Reduce the splitter count in solver/puzzles/ before generating.'
    );
  }

  console.log(`Computing optimal score for ${dateStr} (${config.name})...`);
  const result = await solvePuzzle(config, solverOpts);
  const optimalScore = result.score;
  const optimalMirrors = result.mirrors;
  console.log(`  Optimal score: ${optimalScore} (${optimalMirrors.length} mirrors)`);

  // Convert mirrors from [x, y, type] triples to JSON-serialisable format
  const optimalSolution = optimalMirrors.map(([x, y, type]) => ({ x, y, type }));

  const dirName = ['up', 'right', 'down', 'left'][config.laserDir];

  return {
    date: dateStr,
    gridWidth: config.width,
    gridHeight: config.height,
    laserConfig: {
      x: config.laserX,
      y: config.laserY,
      direction: dirName,
    },
    obstacles: [
      // Plain wall obstacles
      ...config.obstacles.map(([x, y]) => ({ x, y })),
      // Splitters (emitted after walls, matching Python order)
      ...(config.splitters || []).map(([x, y, orientation]) => ({
        x, y, type: 'splitter', orientation,
      })),
    ],
    mirrorsAvailable: config.numMirrors,
    optimalScore,
    optimalSolution,
  };
}

/**
 * Generate and save a level for a specific date.
 */
async function generateForDate(dateStr: string, outputDir: string, solverOpts: SolverOpts) {
  const config = getPuzzleForDate(dateStr);
  const level = await generateLevel(dateStr, config, solverOpts);

  const outputPath = path.join(outputDir, `${dateStr}.json`);
  fs.writeFileSync(outputPath, JSON.stringify(level, null, 2));
  console.log(`  Saved to ${outputPath}`);
}

// ── Date utilities ────────────────────────────────────────────────────────────

function parseDate(str: string): Date {
  const d = new Date(str + 'T00:00:00Z');
  if (isNaN(d.getTime())) throw new Error(`Invalid date: ${str}`);
  return d;
}

function formatDate(d: Date): string {
  return d.toISOString().slice(0, 10);
}

function addDays(d: Date, n: number): Date {
  const r = new Date(d);
  r.setUTCDate(r.getUTCDate() + n);
  return r;
}

// ── CLI ───────────────────────────────────────────────────────────────────────

function parseArgs(argv: string[]) {
  const args = argv.slice(2);
  const opts = {
    date: null as string | null,
    start: null as string | null,
    end: null as string | null,
    beamWidth: 8000,
    workers: os.cpus().length,
  };
  for (let i = 0; i < args.length; i++) {
    const a = args[i];
    if (a === '--date')        opts.date      = args[++i];
    else if (a === '--start')  opts.start     = args[++i];
    else if (a === '--end')    opts.end       = args[++i];
    else if (a === '--beam-width') opts.beamWidth = parseInt(args[++i], 10);
    else if (a === '--workers')    opts.workers   = parseInt(args[++i], 10);
    else { console.error(`Unknown argument: ${a}`); process.exit(1); }
  }
  return opts;
}

async function main() {
  const opts = parseArgs(process.argv);

  // Validate arguments (matches Python generate_levels.py behaviour)
  if (!opts.date && !(opts.start && opts.end)) {
    console.error('Error: must specify either --date or both --start and --end');
    process.exit(1);
  }
  if (opts.date && (opts.start || opts.end)) {
    console.error('Error: cannot use --date with --start/--end');
    process.exit(1);
  }
  if ((opts.start && !opts.end) || (opts.end && !opts.start)) {
    console.error('Error: must specify both --start and --end for range generation');
    process.exit(1);
  }

  const outputDir = path.join(__dirname, 'levels');
  fs.mkdirSync(outputDir, { recursive: true });

  const solverOpts = { beamWidth: opts.beamWidth, workers: opts.workers };

  try {
    if (opts.start && opts.end) {
      // Generate date range
      let current = parseDate(opts.start);
      const end = parseDate(opts.end);
      while (current <= end) {
        await generateForDate(formatDate(current), outputDir, solverOpts);
        current = addDays(current, 1);
      }
    } else {
      // Generate single specific date
      await generateForDate(opts.date!, outputDir, solverOpts);
    }
    console.log('\nDone!');
  } catch (err: any) {
    console.error(`Error: ${err.message}`);
    process.exit(1);
  }
}

main().catch(err => { console.error(err); process.exit(1); });
